### ISSUE TITLE
Is system-console-users sustainable?

### ISSUE THREAD

**Issue by Vincent43** [bus1/dbus-broker#110]
Hi, I'm wondering about new build option `system-console-users` which needs every display manager system user provided as an argument.

If I understand this correctly every distro package  have to provide all known relevant system-users (mostly used by display managers) at build time. Is this option sustainable knowing there are many display managers (gdm,sddm,lightdm, etc.) which can be configured with specific system user account? Does dbus-broker maintainers have to know them all beforehand?

Moreover if someone want to change those users they have to rebuild dbus-broker on their own. If someone want to use display manager unsupported by distro, they have to rebuild dbus-broker again.

Having `system-console-users` as build-time option isn't sustainable in my opinion. Runtime option will be better or some other solution.

Disclaimer: upgrading to v11 broke my display manager.

**Comment by teg** [comment]
Could you open a separate issue about how your display manager broke? If you attach "journalctl -b -u dbus-broker -o verbose" it would help understanding any policy issues.

**Comment by Vincent43** [comment]
I had problems with NetworkManager after dbus-broker update so I assumed this was a cause but now I can't reproduce it. It can be false alarm, I will observe it for next days. Maybe my system is trolling me?
```
NetworkManager[1460]: <error> [1519465033.4844] auth: could not get polkit proxy: Error calling StartServiceByName for org.freedesktop.PolicyKit1: Timeout was reached
NetworkManager[1460]: <warn>  [1519465033.4849] supplicant: failed to acquire wpa_supplicant proxy: Wi-Fi and 802.1x will not be available (Error calling StartServiceByName for fi.w1.wpa_supplicant1: Timeout was reached)
NetworkManager[1460]: <warn>  [1519465033.4979] error requesting auth for org.freedesktop.NetworkManager.enable-disable-network: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4990] error requesting auth for org.freedesktop.NetworkManager.sleep-wake: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.enable-disable-wifi: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.enable-disable-wwan: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.enable-disable-wimax: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.network-control: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.wifi.share.protected: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.wifi.share.open: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4991] error requesting auth for org.freedesktop.NetworkManager.settings.modify.system: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.settings.modify.own: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.settings.modify.hostname: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.settings.modify.global-dns: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.reload: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.checkpoint-rollback: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4992] error requesting auth for org.freedesktop.NetworkManager.enable-disable-statistics: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.4993] error requesting auth for org.freedesktop.NetworkManager.enable-disable-connectivity-check: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.9489] error requesting auth for org.freedesktop.NetworkManager.wifi.share.protected: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.9490] error requesting auth for org.freedesktop.NetworkManager.wifi.share.open: Authorization check failed: invalid DBUS proxy
NetworkManager[1460]: <warn>  [1519465033.9767] error requesting auth for org.freedesktop.NetworkManager.network-control: Authorization check failed: invalid DBUS proxy
```

**Comment by Vincent43** [comment]
Let's forget about this for now. I'll re-open if you find something.

**Comment by dvdhrm** [comment]
This looks like you live-updated dbus-broker without rebooting, right? Did you update anything else at the same time? The NetworkManager error is triggered by a timeout on StartServiceByName(), which is implemented by dbus-broker and triggers a StartUnit() call on the respective systemd unit. In your case, the unavailable service is PolicyKit1, and systemd seems to time-out activating it.

I suspect a restart of NetworkManager would have solves this issue, since it looks like NetworkManager cached the error and invalidated its proxy-object, thus repeating the error over and over.

Regarding the reasons for this error, I am left with guessing. Since I assume it happened during a live-update on a running system, many things can go wrong. Since the system is never atomically updated on most distributions, it might be a simple case of updating dbus-broker/systemd/polkit/networkmanager in any combination, and your error was triggered in an inconsistent state.

This might be systemd losing track of the service activation, networkmanager being restarted while systemd/polkit/dbus-broker is updated, or a plain bug in any of these softwares. Hard to tell.

In any case, if you can trigger this again, I'd gladly look into it further. I failed to reproduce it, though. I think this was just about non-atomic distro updates and solved by a reboot, but you can never be sure.

Anyway, thanks a lot for the resport and testing it!

**Comment by dvdhrm** [comment]
> Having system-console-users as build-time option isn't sustainable in my opinion. Runtime option will be better or some other solution.

Correct!

The option was added for completeness, in case distributions want to opt-in. I would always recommend shipping XML policies instead, which open up the required interfaces for the correct users. For instance, if SDDM or gdm want access to a privileged interface, I recommend shipping an XML policy with each of them, which opens the privileged interface for their user.

Ultimately, something like PolicyKit is always easier to use, but it is up to the interface provider to opt-in here. In some cases this did not happen, so workarounds are needed. The `system-console-users` option is only a last resort, but IMHO a natural extension to our `at-console` handling via `SYSTEMUSERMAX`.

**Comment by Vincent43** [comment]
That happened after reboot near the time after I upgraded dbus-broker to version 11. I think I rebooted it again to see if it was reproducible and it was at that time. Restarting NetworkManager after boot was enough to solve this issue.

After reading about `system-console-users` I thought dbus-broker was the cause and I downgraded it to version 10. The issue was gone after reboot. After few days I recompiled dbus-broker with `-Darray_system-console-users==sddm` and there was no issues after reboot so I thought I confirmed the cause of this and reported it here.

After that to be 100% sure that I can reproduce this I recompiled dbus-broker again without `-Darray_system-console-users==sddm` and was surprised that I can't reproduce this issue anymore. I corrected my report here.

**To sum it up**: The issue seems to be pure coincidental and wasn't related to dbus-broker as I first thought. Sorry for the noise.

**Comment by dvdhrm** [comment]
No problem. Thanks for clarifying!
